### PR TITLE
NIAD-2666: Battery Level Filing Comments should be referenced from DiagnosticReport

### PR DESCRIPTION
**Pull request by MartinWheelerMT**
## What

Having Mapping out the TPP and current adaptor behaviour, we have identified the differences between the mapping.  Discussed this further and it appears that the behaviour should be:

* If a comment note narrative statement appears at the a diagnostic report level, then this should be referenced to in the result references.  This observation is determined to be the filing comment
* If there is no comment not narrative statement at diagnostic report level, the it could appear in one or more nested “battery” results.  In this case a new observation comment should be created to be used as the filing comment and referenced to in the results references.  This observation filing comment should not contain a comment text field.

This is a breaking change as it will be adding new references to the `result` references in a `DiagnosticReport` and the consuming system will need to handle this.

## Why

As part of the comparisons under ["Compare Investigations data across TPP and the PS Adaptor, identifying all bugs and issues with an SME"](https://nhse-dsic.atlassian.net/browse/NIAD-2666), it has been identified that filing comments are being linked to the TGH rather than the Diagnostic Report, so is the issue that they should be linked to the Report.  Under GP Connect rules this would present a filed report as un-filed.

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have updated the [Changelog](/CHANGELOG.md) with details of my change in the UNRELEASED section if this change will affect end users
- [ ] A corresponding change has been made to the [Mapping Documentation repository][mapping-docs]

[mapping-docs]: https://github.com/NHSDigital/patient-switching-adaptors-mapping-documentation

### CODE REVIEW

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/diagnosticreport/DiagnosticReportMapper.java [hunk 1]
```diff
@@ -155,20 +159,65 @@ private List<Reference> getSpecimenReferences(RCMRMT030101UKCompoundStatement co
...
+        var resultReferences = getDirectResultReferences(compoundStatement);
 
-        var resultReferences = compoundStatement.getComponent()
-            .stream()
-            .filter(RCMRMT030101UKComponent02::hasNarrativeStatement)
-            .map(RCMRMT030101UKComponent02::getNarrativeStatement)
-            .filter(narrativeStatement -> !narrativeStatement.getText().contains(LAB_REPORT_COMMENT_TYPE))
-            .map(narrativeStatement -> new Reference(new IdType(ResourceType.Observation.name(), narrativeStatement.getId().getRoot())))
-            .collect(toCollection(ArrayList::new));
+        // also include USER COMMENT or UNKNOWN references from nested BATTERY as per NIAD-2666
```
Comment (adrianclay) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/diagnosticreport/DiagnosticReportMapper.java:164]: Oh gosh, I'm not a huge fan of referencing Jira tickets for useful context, as it's my belief that code lives longer than Jira. Also, this code is open source whereas Jira is closed off.

Is there anything useful in Jira which should be pulled into this comment? Ideally the tests should give the most information.

File: gp2gp-translator/src/integrationTest/resources/e2e-mapping/output-json/PWTP9-output.json [hunk 2]
```diff
@@ -13730,6 +13732,10 @@
         "reference": "Specimen/0DB91FFF-DC9C-4246-A8D0-A3D340A2D822"
       } ],
       "result": [ {
+        "reference": "Observation/FA1B0FA9-C9BC-4399-B506-52D0286DFD06"
+      }, {
+        "reference": "Observation/E47265C2-CE59-4F24-B203-5A73CEB79D84"
+      }, {
```
Comment (adrianclay) [inline gp2gp-translator/src/integrationTest/resources/e2e-mapping/output-json/PWTP9-output.json:13738]: Multuiple batteries here, so multiple references. Which goes against what the GP Connect spec says.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/diagnosticreport/DiagnosticReportMapper.java (hunk 1 above)
Comment (MartinWheelerMT) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/diagnosticreport/DiagnosticReportMapper.java]: removed this as is no longer required

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/diagnosticreport/DiagnosticReportMapper.java [hunk 3]
```diff
@@ -155,20 +159,65 @@ private List<Reference> getSpecimenReferences(RCMRMT030101UKCompoundStatement co
...
+        var batteryResultReferences = getNestedBatteryResultReferences(compoundStatement);
+        resultReferences.addAll(batteryResultReferences);
 
         if (!resultReferences.isEmpty()) {
             diagnosticReport.setResult(resultReferences);
         }
     }
 
+    @NotNull
+    private ArrayList<Reference> getNestedBatteryResultReferences(RCMRMT030101UKCompoundStatement compoundStatement) {
```
[the remark this answers is not in the document]
Comment (MartinWheelerMT) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/diagnosticreport/DiagnosticReportMapper.java]: This has now also been changed as we only use one filing comment note

File: gp2gp-translator/src/integrationTest/resources/e2e-mapping/output-json/PWTP9-output.json (hunk 2 above)
Comment (MartinWheelerMT) [reply gp2gp-translator/src/integrationTest/resources/e2e-mapping/output-json/PWTP9-output.json:13736]: Have addressed this following our conversation with Martin.  There may be multiple batteries, but only one filing comment should be created to indicate the report as filed

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,14 +5,17 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased]
 
+### Added
+* ** Breaking Change **: When Filing comment `NarrativeStatements` are located within the `BATTERY`
```
Comment (adrianclay) [inline CHANGELOG.md:9]: Is this a breaking change? 🤔

**Review by adrianclay — DISMISSED** [review]

**Review by adrianclay — APPROVED** [review]

**Pull request merged**